### ISSUE TITLE
Login page - Exception: Correlation failed

### ISSUE THREAD

**Issue by YuriyTigiev** [dotnet-architecture/eShopOnContainers#1391]
Hello, 

I build and run the solution in VS 2019 Version 16.7.0  

I open the solution by address 
http://docker.for.win.localhost:5100 or http://docker.for.linux.localhost:5100 
When I'm trying to open a login page then I'm getting an error (see below). 
I have registered in the system (one time) and now I'm trying to login to the system. 

Docker use a WSL. 
Docker version 2.3.0.4 

Login page error 
---
An unhandled exception occurred while processing the request.
Exception: Correlation failed.
Unknown location

Exception: An error was encountered while handling the remote login.
Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler<TOptions>.HandleRequestAsync()

Stack Query Cookies Headers Routing
Exception: Correlation failed.

Show raw exception details
Exception: An error was encountered while handling the remote login.
Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler<TOptions>.HandleRequestAsync()
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

Show raw exception details

Login page error 2
--- 
But sometimes I'm getting an error (It happens randomly)
Error
Sorry, there was an error: unauthorized_client
Request Id: 0HM1SIIUVKC9P:00000010

**Comment by Harsh2016** [comment]
I resolved it  after deleting and recreating the IdentityDb.

**Comment by YuriyTigiev** [comment]
> I resolved it after deleting and recreating the IdentityDb.

How to do that?

**Comment by sughosneo** [comment]
Hi @YuriyTigiev, you may want to look into this [unauthorized_client-error-on-login](https://github.com/dotnet-architecture/eShopOnContainers/wiki/unauthorized_client-error-on-login).

Thanks.

**Comment by Harsh2016** [comment]
Correlation Failed resolved by making changes in webmvc/startup.cs

services.Configure<CookiePolicyOptions>(options =>
            {
                // This lambda determines whether user consent for non-essential cookies is needed for a given request.
                **options.CheckConsentNeeded = context => true;
                options.MinimumSameSitePolicy = SameSiteMode.None;**
                
                options.OnAppendCookie = cookieContext =>
                    CheckSameSite(cookieContext.Context, cookieContext.CookieOptions);
                options.OnDeleteCookie = cookieContext =>
                    CheckSameSite(cookieContext.Context, cookieContext.CookieOptions);
            });

            **app.UseCookiePolicy();**   
            app.UseAuthentication();

**Comment by sughosneo** [comment]
Yes, agreed with @Harsh2016, Thanks for sharing. 

`Correlation Failed` issue could be latest version of Chrome browser. Please check out this reference : https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/

Alternatively you can change the chrome browser settings as well if you don't want to change the code. 

![image](https://user-images.githubusercontent.com/13281246/89754202-91bc4200-daf8-11ea-846d-0155c07085a7.png)

@YuriyTigiev, please check if this helps.

**Comment by mvelosop** [comment]
@sughosneo, perhaps a definite solution could be forwarding the login through the API gateway, so everything is served from that address.

**Comment by YuriyTigiev** [comment]
@sughosneo  

Can you use the advice that offered @Harsh2016  and update the solution?

**Comment by sughosneo** [comment]
Hi @YuriyTigiev, there could be few other ways, the resolution of this issue could be looked at. For an e.g. the suggestion by @mvelosop in his comments. So it's under discussion to see which will have less impact but yet a definite fix for all set of deployments.

In the meantime, could you please confirm if chrome settings changes had actually fixed your issue ? or did you have to make code changes in your local ?

Thank you.

**Comment by YuriyTigiev** [comment]
@sughosneo  

@sughosneo  

I have used the "SameSite" option and it works.

**Comment by mvelosop** [comment]
@YuriyTigiev, @Harsh2016, @sughosneo,

SameSite = None, should only work when using HTTPS in recent browsers is that your case?

**Comment by YuriyTigiev** [comment]
In my case, it's work for HTTP. 

For https (https://docker.for.win.localhost:5105/) I'm getting an error
This site can’t provide a secure connectiondocker.for.win.localhost sent an invalid response.
Try running Windows Network Diagnostics.
ERR_SSL_PROTOCOL_ERROR

**Comment by sughosneo** [comment]
@mvelosop - For me also it happens with HTTP. One of the other behavior which I have noticed is that I am only able to reproduce this with chrome browser in MacOS but windows with same major Chrome version, I don't get to see this behavior.

I think I am missing some point. Thought ? I will have to recheck. 

Thanks.

**Comment by YuriyTigiev** [comment]
@sughosneo 

I have a question. Where we can discuss the architecture of the solution?

**Comment by sughosneo** [comment]
Hi @YuriyTigiev, did you get a chance to look into the [Architecture Wiki](https://github.com/dotnet-architecture/eShopOnContainers/wiki/Architecture). We have recently added [Code Flow](https://github.com/dotnet-architecture/eShopOnContainers/wiki/Code-Flow) as well for different component. 

If you have any questions on any specific topics, please feel free to create a new issue. 

Thanks.

**Comment by sughosneo** [comment]
> @mvelosop - For me also it happens with HTTP. One of the other behavior which I have noticed is that I am only able to reproduce this with chrome browser in MacOS but windows with same major Chrome version, I don't get to see this behavior.
> 
> I think I am missing some point. Thought ? I will have to recheck.
> 
> Thanks.

Now, In windows I see the warning like as per below : 

![image](https://user-images.githubusercontent.com/13281246/90031564-00f18c00-dcdb-11ea-99ce-429c86eda169.png)

My chrome build version "Version 84.0.4147.105 (Official Build) (64-bit)"

@YuriyTigiev / @mvelosop  - Could you please share the chrome build version ?

**Comment by YuriyTigiev** [comment]
Windows 10 64 bit 
Chrome - Version 84.0.4147.105 (Official Build) (64-bit)

**Comment by huangyingting** [comment]
do we have any updates on this, I am running docker-compose locally and I am not able to logn even turn off samesite from chrome browser.
https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/
I also followed above URL and implement the code change.

**Comment by sughosneo** [comment]
> do we have any updates on this, I am running docker-compose locally and I am not able to logn even turn off samesite from chrome browser.
> https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/
> I also followed above URL and implement the code change.

Hi @huangyingting , thank you for reaching out. 

We are in process of discussing some architectural changes which should resolve similar issues permanently. 
At this point, the `samesite` setting should handle the `Correlation failed.` error.
Is there any other error message do you see when you click on the login button ?

**Comment by huangyingting** [comment]
> > do we have any updates on this, I am running docker-compose locally and I am not able to logn even turn off samesite from chrome browser.
> > https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/
> > I also followed above URL and implement the code change.
> 
> Hi @huangyingting , thank you for reaching out.
> 
> We are in process of discussing some architectural changes which should resolve similar issues permanently.
> At this point, the `samesite` setting should handle the `Correlation failed.` error.
> Is there any other error message do you see when you click on the login button ?

thank you, I think I found the solution here, I will submit a PR later

**Comment by cuno92** [comment]
> > > do we have any updates on this, I am running docker-compose locally and I am not able to logn even turn off samesite from chrome browser.
> > > https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/
> > > I also followed above URL and implement the code change.
> > 
> > 
> > Hi @huangyingting , thank you for reaching out.
> > We are in process of discussing some architectural changes which should resolve similar issues permanently.
> > At this point, the `samesite` setting should handle the `Correlation failed.` error.
> > Is there any other error message do you see when you click on the login button ?
> 
> thank you, I think I found the solution here, I will submit a PR later

Thanks you, follow by your PR and it fixed this issue on localhost

**Comment by sughosneo** [comment]
By default the `SameSite` setting is set as `Lax`. Tested the application. 
Closing this issue as of now. Will reopen it later if something gets reported on this aspect. 

Thank you.

**Comment by mina5500** [comment]
not working for mw with lax on chrome 87 http

**Comment by mina5500** [comment]
> Correlation Failed resolved by making changes in webmvc/startup.cs
> 
> services.Configure(options =>
> {
> // This lambda determines whether user consent for non-essential cookies is needed for a given request.
> **options.CheckConsentNeeded = context => true; options.MinimumSameSitePolicy = SameSiteMode.None;**
> 
> ```
>             options.OnAppendCookie = cookieContext =>
>                 CheckSameSite(cookieContext.Context, cookieContext.CookieOptions);
>             options.OnDeleteCookie = cookieContext =>
>                 CheckSameSite(cookieContext.Context, cookieContext.CookieOptions);
>         });
> 
>         **app.UseCookiePolicy();**   
>         app.UseAuthentication();
> ```

@Harsh2016 could you please specify exactly where to write this code in webmvc/startup.cs

**Comment by MullerWasHere** [comment]
I'm still getting Correlation Failed while trying to log in on Chrome v87.0.4280.141.
Currently using http://localhost:5100 to access eShopOnContainers, followed #1236 to allow me to do so.
Built via CLI.

Am I missing something?

**Comment by sughosneo** [comment]
Hi @MullerWasHere , thank you for reaching out.
Did you take the latest from the `dev` branch ? This issue has been fixed in PR #1516
Could you please check the `SameSite by default cookies` value in `chrome://flags/` ? It should be as per below :

![image](https://user-images.githubusercontent.com/13281246/105154859-fa8bf980-5b2f-11eb-9387-831331035520.png)

Also, you should be using below urls to access the application. 

```text

Web Status : http://host.docker.internal:5107/
Web MVC :  http://host.docker.internal:5100/
Web SPA :  http://host.docker.internal:5104/

```

Hope this helps !

**Comment by MullerWasHere** [comment]
@sughosneo, accessing through the recommended URLs fixed it for me. Thank you!
